### DISCUSSION TITLE
Adding IAudioEndpointVolume does nothing, are there troubleshooting docs?

### DISCUSSION THREAD

**Discussion by riverar** [microsoft/CsWin32#476] [Q&A] [answered]
Howdy. Haven't touched CsWin32 in a long while so I'm a bit rusty. I added `IAudioEndpointVolume` and `CreateFile` (as a control) to my NativeMethods.txt in a .NET 6 WPF project. The generated PInvoke class contains a definition for `CreateFile` but my interface is MIA. I've verified this is specified in Win32Metadata so not sure where to go from here.

Are there general troubleshooting steps documented somewhere? Logging? Binlog doesn't appear to be helpful in codegen scenarios.

**Comment by FreddyDgh** [comment, marked as answer]
@riverar FYI - I just created a .NET 6 WPF project and tested this, and it seems to work fine for me.  `IAudioEndpointVolume` is being generated by CsWin32 in a file called `Windows.Win32.IAudioEndpointVolume.g.cs`.  I am using the following NuGet packages:

Microsoft.Windows.CsWin32: 0.1.619-beta
Microsoft.Windows.SDK.Win32Metadata: 15.0.22-preview

I'm not sure if there are any official troubleshooting steps, but when I have issues, I do the following and I typically am able to get things working again:
- Update to the latest NuGet packages (both CsWin32 and Win32Metadata)
- Rebuild the project (not just build again, but literally clean and rebuild)
- Check the Build output and see if there are any messages being generated there

Not sure if any of this is helpful to your scenario, but I figured I'd offer the info.

**Reply by AArnott** [reply]
It works for me with that version too. 
Sometimes the API in NativeMethods.txt gets pasted with an invisible character before or after it, causing it to fail the API lookup. Copying the API name from the docs will sometimes do this. Try deleting the entire line and typing the API by hand, or deleting all invisible characters around the api name (including newlines) and then replace the newlines to see if that helps.

**Reply by riverar** [reply]
@FreddyD-GH @AArnott Thanks for taking a look. I created a new project and can repro the issue; doesn't work.

I'll record a video / create steps, but generally I'm just creating a .NET 6 WPF project, adding NuGet references to both packages stated above, then creating a `NativeMethods.txt` with the aforementioned functions/methods, tweaking the project to compile for x86, then finally building.

**Reply by riverar** [reply]
https://user-images.githubusercontent.com/475132/151061115-74c594b7-40b5-4797-b172-fa2bb9bf250b.mp4

Video errata: I also checked for `Windows.Win32.IAudioEndpointVolume`, etc. nothing. Don't have `Windows.Win32.IAudioEndpointVolume.g.cs` on disk either.

**Reply by jnm2** [reply]
> Try deleting the entire line and typing the API by hand, or deleting all invisible characters around the api name (including newlines) and then replace the newlines to see if that helps.

Shouldn't the issue be clearly explained in the error list if this happens?

**Reply by FreddyDgh** [reply]
@riverar I just looked quickly at your video, but it looks like you are trying to find it in the wrong namespace; the full type name that gets generated is `Windows.Win32.Media.Audio.Endpoints.IAudioEndpointVolume` .  Also, if you are not already aware, you can see the generated files under the analyzers node in Visual Studio when using source generators.

**Reply by riverar** [reply]
@FreddyD-GH Thanks, in hindsight I should've matched that up with the win32metadata duh. You're right, it's there!

But interestingly, Visual Studio 2022 IntelliSense doesn't seem to know how to find this interface prior to explicitly qualifying the type. This is unusual and sounds like a source generator quirk/bug?

(In other words, I typed IAudioEndpointVolume, hit `Ctrl-.`, and VS failed to find the interface.)

**Reply by FreddyDgh** [reply]
I have also noticed that Intellisense doesn't seem to work properly for CsWin32. It seems like often I am not able to automatically find namespaces and sometimes there are erroneous red squiggles that I have to ignore. I don't have the same problem with other source generators - it seems to be specific to CsWin32. @AArnott Do you have any idea why Intellisense and CsWin32 are not playing nicely with each other?

**Reply by AArnott** [reply]
Ctrl-T works for me to find the fully-qualified name. I know this technique has worked for a few minor VS releases. In fact in the latest prerelease I'm on, Ctrl-. works too, which is even better.

**Reply by AArnott** [reply]
> Do you have any idea why Intellisense and CsWin32 are not playing nicely with each other?

I don't. I see roslyn features failing _frequently_ with infobars and restarting VS gets it working ... till it fails again. If you ever find a solid repro for it, please use Report a Problem to report it and feel free to link to your feedback ticket here so I can help investigate.

**Reply by riverar** [reply]
Thanks @FreddyD-GH @AArnott, I am unblocked so will resolve this. Knew it had to be something silly on part.

**Reply by riverar** [reply]
> If you ever find a solid repro for it, please use Report a Problem to report it and feel free to link to your feedback ticket here so I can help investigate.

If it helps, my video above reproduces the issue 100% of the time across all machines. I'll create a VSComm issue.

**Reply by AArnott** [reply]
Watching your video, why did you go to the trouble of installing the win32metadata package separately?
The only red squiggles I see in the video seem to have been temporary and/or when you were crafting in invalid namespace or type name. I'm not sure I see the problem that you'd file on VSComm.

**Reply by riverar** [reply]
@AArnott Is my understanding I can target lower/higher versions of metadata by explicitly adding the package.

re: vscomm, I was trying to point out that the _workflow_ in the video reproduces the `CTRL+.` issue every time (which is what ultimately led me here for support to begin with hah!) Apologies for the crummy video/confusion.

**Reply by AArnott** [reply]
I see. Well, Ctrl+. works in the latest prereleases of VS (internal at least) for CsWin32:
![image](https://user-images.githubusercontent.com/3548/151095929-f1111cb9-54a5-4141-98a0-97e20e284d8b.png)

So you may not need to file anything then.

> Is my understanding I can target lower/higher versions of metadata by explicitly adding the package.

True, within the compatibility limits of CsWin32. For example the latest version of CsWin32 on nuget.org is incompatible with the 17.0 version of the metadata. https://github.com/microsoft/CsWin32/pull/489 recently took the update and had to make several changes.
